lang/funcs/collection.go

```diff
@@ -512,9 +512,14 @@ var LookupFunc = function.New(&function.Spec{
 		}
 
 		ty := args[0].Type()
-		key := args[1].AsString()
+
 		switch {
 		case ty.IsObjectType():
+			if !args[1].IsKnown() {
+				return args[2].Type(), nil
```

Comment:
Doing this check in here does seem right but I don't think returning the "default" value type here is the right fallback for a couple reasons:
- There might not be a default value (it's optional, and so I think this might cause an out-of-bounds panic)
- Once we _do_ know the key (at apply time) we might find that it matches one of the attributes in `inputMap` after all, and so that'll be the type of our result.

With both of those things in mind, I think the best we can do in this scenario is `return cty.DynamicPseudoType, nil`, which effectively means "I don't know yet what the result type will be". That's an okay response when arguments are unknown: it'll just mean we won't catch a type error here during the plan operation. We'll re-run the function again at apply time when we've learned more information, and thus catch any name error there, as usual for unknown values.

Reply:
That makes sense. I made that change - PTAL.